lib/qr_code/masking.ex

```diff
@@ -35,6 +35,134 @@ defmodule QRCode.Masking do
     end)
   end
 
+  def penalty_1(matrix) do
+    row_pen =
+      matrix
+      |> compute_penalty_1()
+
+    col_pen =
+      matrix
+      |> Matrix.transpose()
+      |> compute_penalty_1()
+
+    row_pen + col_pen
+  end
+
+  def penalty_2(matrix) do
+    matrix
+    |> compute_penalty_2()
+  end
+
+  def penalty_3(matrix) do
+    row_pen =
+      matrix
+      |> compute_penalty_3()
+
+    col_pen =
+      matrix
+      |> Matrix.transpose()
+      |> compute_penalty_3()
+
+    row_pen + col_pen
+  end
+
+  def penalty_4(matrix) do
+    {rs, cs} = Matrix.size(matrix)
+
+    dark_modules =
+      matrix
+      |> Enum.reduce(0, fn row, acc -> Enum.sum(row) + acc end)
+
+    percent_of_dark = Kernel.floor(dark_modules * 100 / (rs * cs))
+
+    reminder =
+      percent_of_dark
+      |> Kernel.rem(5)
+
+    Kernel.min(
+      Kernel.abs(percent_of_dark - reminder - 50) / 5,
+      Kernel.abs(percent_of_dark - reminder - 45) / 5
+    ) * 10
+  end
+
+  defp compute_penalty_1(matrix) do
+    matrix
+    |> Enum.reduce(0, fn [h | _] = row, acc ->
+      row
+      |> Enum.reduce({h, 0, acc}, &evaluate_cond_1/2)
+      |> (fn {__selected, _sum, val} -> val end).()
+    end)
+  end
+
+  defp evaluate_cond_1(val, {selected, sum, acc}) when val == selected and sum < 4 do
+    {val, sum + 1, acc}
+  end
+
+  defp evaluate_cond_1(val, {selected, 4, acc}) when val == selected do
+    {val, 5, acc + 3}
+  end
+
+  defp evaluate_cond_1(val, {selected, sum, acc}) when val == selected and sum > 4 do
+    {val, sum + 1, acc + 1}
+  end
+
+  defp evaluate_cond_1(val, {_val, _sum, acc}) do
+    {val, 1, acc}
+  end
+
+  defp compute_penalty_2(rows, acc \\ 0)
+
+  defp compute_penalty_2(rows, acc) when length(rows) == 1 do
+    acc
+  end
+
+  defp compute_penalty_2([row1, row2 | rows], acc) do
+    acc_row =
+      row1
+      |> map2(row2, fn v1, v2 -> v1 + v2 end)
```

Comment:
Why use new `map2` function instead `Enum.zip/2` and lambda function `fn {v1, v2} -> v1 + v2 end` in `Enum.map`?

Reply:
Hmm I really don't know now :-) why I used `map2` approach instead of `Enum.zip`...